packages/file_selector/file_selector_ios/lib/file_selector_ios.dart

```diff
@@ -0,0 +1,69 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:file_selector_platform_interface/file_selector_platform_interface.dart';
+import 'package:flutter/foundation.dart' show visibleForTesting;
+import 'package:flutter/services.dart';
+
+import 'src/messages.g.dart';
+
+const MethodChannel _channel =
+    MethodChannel('plugins.flutter.io/file_selector_ios');
+
+/// An implementation of [FileSelectorPlatform] for iOS.
+class FileSelectorIOS extends FileSelectorPlatform {
+  /// The MethodChannel that is being used by this implementation of the plugin.
+  @visibleForTesting
+  MethodChannel get channel => _channel;
+
+  final FileSelectorApi _hostApi = FileSelectorApi();
+
+  /// Registers the iOS implementation.
+  static void registerWith() {
+    FileSelectorPlatform.instance = FileSelectorIOS();
+  }
+
+  @override
+  Future<XFile?> openFile({
+    List<XTypeGroup>? acceptedTypeGroups,
+    String? initialDirectory,
+    String? confirmButtonText,
+  }) async {
+    final List<String>? path = (await _hostApi.openFile(FileSelectorConfig(
+            utis: _allowedUtiListFromTypeGroups(acceptedTypeGroups),
+            allowMultiSelection: false)))
+        ?.cast<String>();
+    return path == null ? null : XFile(path.first);
+  }
+
+  @override
+  Future<List<XFile>> openFiles({
+    List<XTypeGroup>? acceptedTypeGroups,
+    String? initialDirectory,
+    String? confirmButtonText,
+  }) async {
+    final List<String>? pathList = (await _hostApi.openFile(FileSelectorConfig(
+            utis: _allowedUtiListFromTypeGroups(acceptedTypeGroups),
+            allowMultiSelection: true)))
+        ?.cast<String>();
+    return pathList?.map((String path) => XFile(path)).toList() ?? <XFile>[];
+  }
+
+  // Converts the type group list into a list of all allowed UTIs, since
+  // iOS doesn't support filter groups.
+  List<String>? _allowedUtiListFromTypeGroups(List<XTypeGroup>? typeGroups) {
+    if (typeGroups == null || typeGroups.isEmpty) {
+      return null;
+    }
+    final List<String> allowedUTIs = <String>[];
+    for (final XTypeGroup typeGroup in typeGroups) {
+      // If any group allows everything, no filtering should be done.
```

Comment:
There's an open question here about what we should do if the group isn't empty, but the types supported by the current platform is empty. This will actually give people some surprising results (e.g., they think they are restricting to text files because they passed text extensions and mime types, but then it will do no restriction on iOS).

We don't necessarily need to address here since we have the same issue on other platforms (and should maybe address it later on all of them), but this is an extreme version since the most obscure type is the only one iOS supports. I think it might be better to check whether it's completely empty, and if not, throw an `ArgumentError`. (Which ideally we would do with a small precursor PR to add an `isEmpty` to `XTypeGroup` so it's not fragile to later filter type additions.) That's the change I've been considering for the others—as a breaking change to the plugin, unfortunately—but we could do it here first without it being breaking.

Reply:
I'm fixing this issue in https://github.com/flutter/plugins/pull/6120, so we should update this PR to follow the structure there (which is the approach I describe in the comment above)

Reply:
Done.